[css-scroll-snap] Skip renderers with no element in updateSnapOffsetsForScrollableArea

nmoucht opened the pull request:
#### 2f88930ccc922eb8567c8fe24a43ef945bea0127
<pre>
[css-scroll-snap] Skip renderers with no element in updateSnapOffsetsForScrollableArea
<a href="https://bugs.webkit.org/show_bug.cgi?id=263093">https://bugs.webkit.org/show_bug.cgi?id=263093</a>
<a href="https://rdar.apple.com/111579277">rdar://111579277</a>

Reviewed by Simon Fraser.

We should skip renderers in updateSnapOffsetsForScrollableArea if they have no related element.
For generated content we do generate an associated element as well, so the only time a renderer&apos;s
element is null is for anonymous renderers (as specified in RenderElement.h as well), which we
shouldn&apos;t snap to.

* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):

Canonical link: <a href="https://commits.webkit.org/275523@main">https://commits.webkit.org/275523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36d16755a93237a1dd7adb607b5b1937a37e9204

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42064 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44651 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38177 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18410 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34838 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42638 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18003 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36216 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15770 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15677 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46082 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37576 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16880 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13873 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40034 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/36520 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18559 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5654 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
```diff
@@ -388,7 +388,8 @@ void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const Re
         auto isFocused = child.element() ? focusedElement == child.element() : false;
-        auto identifier = child.element() ? child.element()->identifier() : ObjectIdentifier<ElementIdentifierType>(0);
+        auto identifier = child.element()->identifier();
+        ALWAYS_LOG_WITH_STREAM(stream << "updateSnapOffsetsForScrollableArea child renderer: "  << child << " child element: " << child.element() << " identifier: " << identifier);
```
This should be `LOG_WITH_STREAM(ScrollSnap, ...)`.

nmoucht replied:
Removed.

heycam commented on Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
```diff
@@ -347,7 +347,7 @@ void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const Re
     LOG_WITH_STREAM(ScrollSnap, stream << "Computing scroll snap offsets for " << scrollableArea << " in snap port " << scrollSnapPort);
     for (auto& child : boxesWithScrollSnapPositions) {
-        if (child.enclosingScrollableContainerForSnapping() != &scrollingElementBox)
+        if (child.enclosingScrollableContainerForSnapping() != &scrollingElementBox || !child.element())
```
Aside from resolving the crash, can you explain why it's correct that we skip this work when `child.element()` is null? It could be null when the box is for generated content like a `::before`. Is it right that we skip updating snap offsets for generated content? (Do we support scroll snapping to generated content? The example at https://drafts.csswg.org/css-scroll-snap-1/#re-snap suggests that we should, but I don't know if we do.)

heycam replied:
I might be wrong, and we do return a non-null element here for generated content. So maybe it is for anonymous content? In which case it is fine? Either way, please do explain in the commit message why it is OK to skip in this case.

nmoucht replied:
Hmm yeah I haven't enumerated all of the cases where we have a renderer but not an element, so I'm not sure if it's ok in all cases. I'll try to make some tests around this.

nmoucht replied:
Confirmed that we do generate an element in this case, so it would only be null for anonymous renderers.

smfr replied:
Doesn't `renderer->element()` return the same thing for a pseudo's render and the element's normal renderer? If so, how can you distinguish between snapping to the element, and snapping to the pseudo?

nmoucht replied:
No renderer->element() returns a generated node with type pseudo, with its own unique identifier from the element's renderer.

smfr replied:
It would be good to add a test case that snaps to both elements and their `::before` and `::after` so we prove that those are individual snap targets.

nmoucht replied:
Unfortunately I don't think there's a great way to force snapping to both `::before` and `::after` elements, since the behavior for choosing between them is essentially random (in `chooseBoxToResnapTo` we just choose the first element in snappedBoxes).

smfr commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-to-visible-areas-both-psuedo.html:
```diff
@@ -0,0 +1,92 @@
+<!DOCTYPE html>
+<title>
+  Snap to a visible area only even when there is a closer snap point for an area
```
Is this title (and the text below) accurate for what's being tested here? It looks like this was copied from another test.

nmoucht replied:
Yes this is equivalent to the other test except the two snap areas are pseudo elements. Added this at the end to make it clear.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/snap-to-visible-areas-both-psuedo.html:
nit: psuedo typo in file name

nmoucht replied:
Fixed.

nmoucht committed:
[css-scroll-snap] Skip renderers with no element in updateSnapOffsetsForScrollableArea
https://bugs.webkit.org/show_bug.cgi?id=263093
rdar://111579277

Reviewed by Simon Fraser.

We should skip renderers in updateSnapOffsetsForScrollableArea if they have no related element.
For generated content we do generate an associated element as well, so the only time a renderer's
element is null is for anonymous renderers (as specified in RenderElement.h as well), which we
shouldn't snap to.

* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):

Canonical link: https://commits.webkit.org/275523@main
